config/github/src/main/kotlin/GitHubConfigFileCache.kt

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (C) 2024 The ORT Server Authors (See <https://github.com/eclipse-apoapsis/ort-server/blob/main/NOTICE>)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.eclipse.apoapsis.ortserver.config.github
+
+import io.ktor.utils.io.ByteReadChannel
+import io.ktor.utils.io.core.isEmpty
+import io.ktor.utils.io.core.readByteBuffer
+
+import java.io.File
+import java.io.FileInputStream
+import java.io.InputStream
+import java.io.RandomAccessFile
+import java.nio.channels.FileLock
+
+import kotlin.time.Duration
+
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.withContext
+
+import org.slf4j.LoggerFactory
+
+private val logger = LoggerFactory.getLogger(GitHubConfigFileCache::class.java)
+
+/**
+ * The path under which files downloaded from GitHub are stored in the cache. This is a subdirectory of the folder for
+ * the specific revision.
+ */
+private const val FILES_PATH = "tree"
+
+/**
+ * A class implementing a file cache for data fetched from the GitHub configuration repository.
+ *
+ * Caching of content downloaded from GitHub should be rather effective, as the revision can be used as a criterion to
+ * construct cache keys. It is guaranteed that the data of a specific revision does not change. By applying caching,
+ * the number of requests sent to the GitHub REST API can be reduced significantly. This is especially important since
+ * there is a rate limit in place (which is by default 5000 API requests per hour).
+ *
+ * The implementation has to deal with some race conditions. Since the cache is used by multiple JVMs, it has to be
+ * ensured that concurrent access does not cause any problems. For instance, if a file which is not yet present in the
+ * cache is requested by two JVM instances at the same time, it should be requested from GitHub only once, and it can
+ * be read only after it has been fully downloaded. The implementation achieves this by using exclusive and shared
+ * locks on the accessed files.
+ *
+ * Note: This implementation focuses on the specific requirements of [GitHubConfigFileProvider]. It may be possible to
+ * extract common functionality of a file-based cache for more generic use cases later.
+ */
+class GitHubConfigFileCache(
```

Comment:
Another point I don't see mentioned: this cache is only growing and entries never seem to be removed.
Shouldn't it be mentioned in the documentation ?
Do you plan on using a "real" caching library with some eviction mechanism such as, from the top of my head, EhCache ?

Reply:
I am going to add clean-up logic in a follow-up PR. I don't plan to use a caching library here. I think the use case is rather special, since a fix revision guarantees that files will not become invalid in the cache. Therefore, the standard eviction strategies are probably not suitable.

Reply:
I was thinking more like a time-based eviction policy, e.g. "if this file was not accessed since x months, every it from the cache".

Reply:
Eviction will apply to whole revisions. When the `main` branch moves forward, older revisions are typically no longer used.